### ISSUE TITLE
Limma meta analysis

### ISSUE THREAD

**Issue by aidarripoll** [GabrielHoffman/misc_vp#1]
Hi!

I see you use **metafor::rma()** to run a meta-analysis via linear (mixed-effects) models in [limma_meta_analysis.R](https://github.com/GabrielHoffman/misc_vp/blob/master/limma_meta_analysis.R). I noticed you are using as input the MArrayLM2 object from eBayes() -after running dream()-. In my case, I'm also running dream() since I have a mixed model, but when running dream() I'm already passing as `L`  parameter the contrast matrix I obtained from: `L = getContrast(vobjDream_obj, form_model, metadata_obj, comb[[2]])`, where:
* vobjDream = voomWithDreamWeights(ge_dge, form, metadata)
* form_model #~Gender + Age + (1 | date)
* metadata_obj #metadata object
* comb[[2]]) #GenderF

Then, my  MArrayLM2 object from dream() will be: `fit_dream = dream(vobjDream_obj, form_model, metadata_obj, L)`.

So, I'm wondering which of the two MArrayLM2 objects I should use for the meta-analysis:
1) the MArrayLM2 object resulting directly from dream() --> `fit_dream`
2) the MArrayLM2 object resulting from eBayes() on the  MArrayLM2 dream object --> `fit_eBayes=eBayes(fit_dream)`

Also, I guess I should use the parameter `method="REML"` in **metafor::rma()**, right?

Thanks a lot!
Aida

**Comment by GabrielHoffman** [comment]
Hi Aida, 
1) Input into meta-analysis:
I recommend `dream() -> eBayes() -> run_meta_analysis()`.  ` run_meta_analysis()` uses `topTable()`  which can work directly on `dream` output, but also after `eBayes` is applied.  But note that using the result of `dream` directly **does not** use empirical Bayes moderation; `eBayes` must be used explicitly.

2) Meta-analysis.
You can either use a fixed-effect approach to test if the coefficients from multiple studies have a shared non-zero value.  Or you can use a random effects approach to test if there is heterogeneity in effect size.  A random effect (i.e. `REML`) is the default here.

Gabriel

**Comment by aidarripoll** [comment]
Ok, I see. However, when trying to run `eBayes` on the dream() MArrayLM2 object I found that error:
`> fit_e <- eBayes(fit)`
*Warning message:
In eBayes(fit) :
  Empirical Bayes moderated test is no longer supported for dream analysis
Returning original results for use downstream*

Could this issue be related to the variancePartition or metafor version? See attached the `sessionInfo()` content:
`> sessionInfo()`
R version 4.0.0 (2020-04-24)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: SUSE Linux Enterprise Server 12 SP2

Matrix products: default
BLAS:   /gpfs/apps/MN4/R/4.0.0/INTEL/lib64/R/lib/libRblas.so
LAPACK: /gpfs/apps/MN4/R/4.0.0/INTEL/lib64/R/lib/libRlapack.so

locale:
[1] C

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
 [1] variancePartition_1.20.0 Biobase_2.50.0           BiocGenerics_0.36.0
 [4] scales_1.1.1             BiocParallel_1.24.1      ggplot2_3.3.5
 [7] edgeR_3.32.0             limma_3.46.0             Matrix.utils_0.9.8
[10] stringr_1.4.0            stringi_1.5.3            reshape2_1.4.4
[13] tibble_3.1.1             tidyr_1.1.3              dplyr_1.0.5
[16] plyr_1.8.6               lme4_1.1-26              Matrix_1.2-18
[19] data.table_1.13.4        optparse_1.6.6

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5          locfit_1.5-9.4      lattice_0.20-41
 [4] snow_0.4-3          prettyunits_1.1.1   gtools_3.8.2
 [7] assertthat_0.2.1    foreach_1.5.1       utf8_1.1.4
[10] R6_2.5.0            pillar_1.6.0        gplots_3.1.1
[13] rlang_0.4.10        progress_1.2.2      minqa_1.2.4
[16] nloptr_1.2.2.2      grr_0.9.5           splines_4.0.0
[19] statmod_1.4.35      munsell_0.5.0       numDeriv_2016.8-1.1
[22] compiler_4.0.0      pkgconfig_2.0.3     lmerTest_3.1-3
[25] tidyselect_1.1.0    codetools_0.2-18    fansi_0.4.1
[28] crayon_1.4.1        withr_2.4.2         MASS_7.3-53
[31] bitops_1.0-6        grid_4.0.0          nlme_3.1-151
[34] gtable_0.3.0        lifecycle_1.0.0     DBI_1.1.0
[37] magrittr_2.0.1      KernSmooth_2.23-18  doParallel_1.0.16
[40] getopt_1.20.3       colorRamps_2.3.1    ellipsis_0.3.1
[43] generics_0.1.0      vctrs_0.3.7         boot_1.3-25
[46] iterators_1.0.13    tools_4.0.0         glue_1.4.2
[49] purrr_0.3.4         hms_1.0.0           pbkrtest_0.4-8.6
[52] colorspace_2.0-0    caTools_1.18.0

Thanks for your time!

Aida

**Comment by GabrielHoffman** [comment]
`variancePartition_1.20.0` and `R 4.0.0` is outdated.  You should upgrade to R 4.2.0 or at least 4.1.2, and reinstall `variancePartition`.

Also check out updated docs here: https://diseaseneurogenomics.github.io/variancePartition/

**Comment by aidarripoll** [comment]
Ok, now it works fine!Input into meta-analysis:
I recommend dream() -> eBayes() -> run_meta_analysis(). run_meta_analysis() uses topTable() which can work directly on dream output, but also after eBayes is applied. But note that using the result of dream directly does not use empirical Bayes moderation; eBayes must be used explicitly.

**Comment by aidarripoll** [comment]
Ok, now it works fine!

Also, in case we are interested only in the DE results from a specific dataset, should we also apply eBayes() on the dream() output before applying topTable()? 

Thanks,
Aida

**Comment by aidarripoll** [comment]
Also, now I'm getting the following warning:

`fit <- limma::eBayes(fit)`
Warning message:
In .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  :
  Estimation of var.prior failed - set to default value

Then, these are the outputs of `topTable()` on the dream() -`fit_Toptable`- and dream() + eBayes() -`fit_Toptable.eBayes`- objects:

`fit_topTable <- topTable(fit, coef=c('L1'), number=length(fit$F.p.value))`
`fit_topTable.eBayes <- topTable(fit_eBayes, coef=c('L1'), number=length(fit$F.p.value))`

`head(fit_topTable)`
```r
           logFC  AveExpr         t      P.Value    adj.P.Val     z.std
XIST    5.688523 4.423999  38.53945 1.786650e-44 1.518410e-40  13.99030
UTY    -4.944338 1.414386 -36.87090 1.896590e-44 1.518410e-40 -13.98605
RPS4Y1 -5.615147 6.941330 -34.21092 4.597130e-42 2.453641e-38 -13.58990
TTTY15 -5.504349 1.932401 -31.67274 5.735347e-41 2.295859e-37 -13.40392
KDM5D  -5.195283 1.752462 -30.38238 1.227727e-39 3.931673e-36 -13.17466
EIF1AY -5.229449 3.213488 -30.49558 1.992235e-39 5.316612e-36 -13.13808
```

`head(fit_topTable.eBayes)`
```r
           logFC  AveExpr         t      P.Value    adj.P.Val        B     z.std
UTY    -4.944338 1.414386 -36.87066 1.290319e-45 2.066059e-41 81.03055 -14.17598
XIST    5.688523 4.423999  37.48991 5.633184e-45 4.509927e-41 62.93326  14.07215
RPS4Y1 -5.615147 6.941330 -34.27134 3.266013e-43 1.743180e-39 73.70357 -13.78211
TTTY15 -5.504349 1.932401 -31.80321 4.322176e-42 1.730167e-38 73.25937 -13.59441
KDM5D  -5.195283 1.752462 -30.54523 9.324649e-41 2.986126e-37 72.30075 -13.36781
EIF1AY -5.229449 3.213488 -30.30508 3.025926e-40 8.075187e-37 64.08135 -13.27995
```
Any suggestion?

Aida

**Comment by GabrielHoffman** [comment]
If you run `dream()`, you can run `eBayes()` afterward on the result.  If your formula only includes fixed effects, then running  `eBayes()` automatically runs `limma::eBayes()` in the backend.  If you specified random effects, then  `eBayes()` accounts for the random effects in the backend.  This is all handled behind the scenes.

Running `limma::eBayes()` "manually" on the result of `dream()` gives an error since the `limma` version  is not designed for random effects.

 In summary, I recommend:

```r
vobj = voomWithDreamWeights( geneExpr, form, metadata )
fit = dream( vobj, form, metadata )
fit = eBayes(fit)
topTable(fit, coef=...)
```

Gabriel
